// common-fields/index.js
var sequelize = require('sequelize');

module.exports = {

  PRIMARY_UUID:function(){
    return {
      type:sequelize.UUID,
      defaultValue: sequelize.UUIDV4,
      primaryKey: true
    }
  },

  PRIMARY_ID:function(){
    return {
      type: sequelize.INTEGER,
      default: 0,
      autoIncrement: true,
      primaryKey: true
    }
  },

  ACTIVE:function(){
    return {
      type:sequelize.BOOLEAN,
      defaultValue:true,
      allowNull:false
    }
  },

  INTEGER:function(defaultValue, allowNull){
    return {
      type:sequelize.INTEGER,
      defaultValue:defaultValue,
      allowNull:allowNull
    }
  },

  ENUM : function(values, defaultValue){
    return {
      type:sequelize.ENUM,
      values:values,
      defaultValue:defaultValue
    }
  },

  STRING : function(max){
    var str = { type:sequelize.TEXT };
    if(typeof max !== void 0) str.validate = max;
    return str;
  },
  STRING_10:  function(){ return this.STRING(10);  },
  STRING_20:  function(){ return this.STRING(20);  },
  STRING_40:  function(){ return this.STRING(40);  },
  STRING_60:  function(){ return this.STRING(60);  },
  STRING_80:  function(){ return this.STRING(80);  },
  STRING_120: function(){ return this.STRING(120); },
  STRING_255: function(){ return this.STRING(255); }

};
