// bomb/index.js

var _ = require('lodash');

var bomb = function(message, code, name, data, debug){
  this.message = message;
  if(code)  this.code = code;
  if(name)  this.name = name;
  if(data)  this.data = data;
  if(debug) this.debug = debug;
};
bomb.prototype.toError = function(){
  var err = new Error(this.message);
  if(this.code)   err.code = this.code;
  if(this.name)   err.name = this.name;
  if(this.data)   err.data = this.data;
  if(this.debug)  err.debug = this.debug;
  return err;
};
bomb.prototype.explode = function(){
  throw this.toError();
};

// create error
var boom = function(message, code, name, data, debug){
  return new bomb(message, code, name, data, debug);
};
var mergeErr = function(err, message, code, name, data, debug){
  var merged = boom('You are not authorized to access this data.', 401, 'Not Authorized');
  if(message) merged.message = message;
  if(code) merged.code = code;
  if(name) merged.name = name;
  if(data) merged.data = data;
  if(debug) merged.debug = debug;
  return merged;
};


// EXPORTS
exports.theBomb = bomb;
exports.boom = boom;


// COMMON ERRORS
exports.UNAUTHORIZED = function(message, code, name, data, debug){
  var err = boom('You are not authorized to access this data.', 401, 'Not Authorized');
  return mergeErr(err, message, code, name, data, debug);
};
exports.BAD_REQUEST = function(message, code, name, data, debug){
  var err = boom('Invalid Request.', 400, 'Bad Request');
  return mergeErr(err, message, code, name, data, debug);
};
exports.NOT_FOUND = function(message, code, name, data, debug){
  var err = boom('The data you requested could not be found', 401, 'Not Found');
  return mergeErr(err, message, code, name, data, debug);
};
